Shibboleth Optional Upgrade and Single Journal Fix

lmeerwood opened the pull request:
This pull request re-enables the use of the Optional setting for Shibboleth login. It also fixes the issue where Single Journal instances would not be able to enable the Shibboleth plugin.

I would appreciate someone else testing this as most of my changes have been copied of code from other parts of the OJS/PKP-lib project. I've tested everything I can think of and everything appears to work well.

Cheers.

Commit:
Add optional setting to form

Commit:
Initial changes to allow optional shibboleth login

Commit:
Make 'Institutional Login' button match rest of theme

Commit:
Added ability to customize Shibboleth labelling on login.
Fixed issue with Shibboleth not working on single journal instances.
Removed some dead code.

lmeerwood commented:
Will fix issue #3 if merged.

ctgraham commented on ShibbolethAuthPlugin.inc.php:
```diff
@@ -30,22 +30,37 @@ class ShibbolethAuthPlugin extends GenericPlugin {
 		parent::__construct();
 		$this->_contextId = $this->getCurrentContextId();
 		$this->_globallyEnabled = $this->getSetting(0, 'enabled');
+                $this->_singleContext = true;
```
I wonder if this is needed, given that this is a "sitewide" plugin (c.f. [version.xml](https://github.com/pkp/shibboleth/blob/5a57107bcc3dc8059c480bb96b451333d71a19c7/version.xml#L19) and [ShibbolethAuthPlugin.inc.php](https://github.com/pkp/shibboleth/blob/5a57107bcc3dc8059c480bb96b451333d71a19c7/ShibbolethAuthPlugin.inc.php#L79-L84).  That is, you wouldn't want to enable it for just one journal in a site, right?  The fix for the single journal install might be easier and clearer as forcing `$_globallyEnabled` and ensuring settings are always written to `CONTEXT_SITE`.

lmeerwood replied:
Oh ok, I'll give that a go and see. I followed the changes from [this ojs_ldap_plugin](https://github.com/shemgp/ojs_ldap_plugin/pull/3/commits/1579dfe6ff07f1fc57e27e4961b5f78023da36a4). I'll have a look at the documentation for `CONTEXT_SITE`.

ctgraham replied:
Interestingly, the ojs_ldap_plugin also describes itself as sitewide.

Generally, we pass a context id (that is, "which journal does this apply to?") into the constructor and functions, but since this is a site plugin, we could just ignore the passed in context id and force everything to the `CONTEXT_SITE` constant.

On the other hand, this then raises the question of what happens for folks who already have data assigned within a specific context?  We'd either need to migrate the data for them (probably a bad idea), or require that they re-do their settings after the upgrade (probably a slightly less bad idea).

Or could there actually be a use case out there where different journals would have different Shib configurations, or LDAP configurations?

lmeerwood replied:
I personally cannot see the use for seperate journals having different authentication methods. The reason I added this change in the first place was for OJS instances that only had single journals. They had issue enabling Shibboleth because it wouldn't let them enable it on their journals, and because they only have one journal they don't have access to the site wide plugins.

ctgraham commented on ShibbolethSettingsForm.inc.php:
```diff
@@ -186,6 +194,30 @@ function execute() {
+		$this->_plugin->updateSetting(
+			$this->_contextId,
+			'shibbolethOptionalTitle',
+			trim($this->getData('shibbolethOptionalTitle'), "\"\';"),
```
Did this `trim` prove necessary in practice, or was it copied from elsewhere?  It seems odd that we would need it.

lmeerwood replied:
This was another copy. I'll remove it and see how it goes.

ctgraham commented on locale/en_US/locale.xml:
```diff
@@ -20,26 +20,33 @@
+        <message key="plugins.generic.shibboleth.manager.settings.shibbolethWayfUrlRequired">Please enter the path of the Shibboleth wayfinder service on this server. (e.g. Shibboleth.sso/Login)</message>
 
-	<message key="plugins.generic.shibboleth.manager.settings.shibbolethHeaderUin">Header with Shibboleth UIN</message>
+	<message key="plugins.generic.shibboleth.manager.settings.shibbolethHeaderUin">Server Environment Variable ($_SERVER) with Shibboleth UIN</message>
```
The new instruction seems a bit confusing to me, but obviously the old instruction seemed confusing to you.  What we want to know here (and elsewhere) is what HTTP Header contains the data value, specifically in how that HTTP Header is expressed from a system environment variable as represented in the $_SERVER superglobal.  Perhaps:
> HTTP Header ($_SERVER environment variable key) ....

lmeerwood replied:
I like that compromise.

ctgraham commented on pages/ShibbolethHandler.inc.php:
```diff
@@ -100,6 +162,10 @@ function shibLogin($args, $request) {
 			'shibbolethHeaderEmail'
 		);
+                
+		foreach($_SERVER as $key => $value){
```
Looks like some leftover debugging code.

ctgraham commented on pages/ShibbolethHandler.inc.php:
```diff
@@ -242,7 +361,8 @@ function _getPlugin() {
 	 * @param $user User
 	 */
-	function _checkAdminStatus($user) {
+        function _checkAdminStatus($user) {
```
Unintended whitespace change?

ctgraham replied:
This is still off, but non-critical.

davidjb replied:
Fixed in 3540812

ctgraham commented on pages/ShibbolethHandler.inc.php:
```diff
@@ -432,7 +562,6 @@ function _shibbolethRedirect($request) {
 		);
-		$shibUrl = $wayfUrl . '?target=' . $shibLoginUrl;
-		return $request->redirectUrl($shibUrl);
+		return "/" . $wayfUrl . '?target=' . $shibLoginUrl;
```
Why force the leading slash here?  Shouldn't it be up to the admin to configure an absolute or relative path?

lmeerwood replied:
I had an issue with single context journals. For example, in my test single context site it would set the redirect to `https://example.com/journalName/Shibboleth.sso/Login` when it was supposed to be just `https://example.com/Shibboleth.ss/Login`.

ctgraham replied:
But if the admin entered "/Shibboleth.sso/Login" rather than "Shibboleth.sso/Login", it would still work, true?

lmeerwood replied:
Hmm true. I'll remove it, but maybe in some documentation somewhere it should be mentioned that you will probably need to start your auth url with a leading slash if you are running a single journal instance?

ctgraham replied:
Or, more specifically, that the path entered will be treated as a relative path unless it starts with a leading slash?

ctgraham replied:
We didn't end up clarifying the documentation here, to my reading.

davidjb replied:
I've pushed a change for this in ff1ce2f6854b8d93f5a06256bd712d26a3ff53e5 - explaining that the WAYF setting can be either a URL or a path, with examples.  I agree with your comments @ctgraham, it's up to admin to enter the value that works for them, be it absolute or relative and shouldn't make any assumptions on what they need.

ctgraham commented on templates/optionalShibbolethLogin.tpl:
```diff
@@ -0,0 +1,116 @@
+{**
```
Another approach rather than replicating a modified template here would be to use a filter to modify the template as it is being rendered.  For example: orcidPlugin
https://github.com/pkp/orcidProfile/blob/e271e8801879bfce64e045f0110f1b8b8094b6bd/OrcidProfilePlugin.inc.php#L57-L58
https://github.com/pkp/orcidProfile/blob/e271e8801879bfce64e045f0110f1b8b8094b6bd/OrcidProfilePlugin.inc.php#L258-L262
https://github.com/pkp/orcidProfile/blob/e271e8801879bfce64e045f0110f1b8b8094b6bd/OrcidProfilePlugin.inc.php#L325-L354

This would allow multiple plugins to modify the login page, presenting additional login options.

lmeerwood replied:
That's amazing. I didn't know that was a thing. I'll give that a shot.

ctgraham replied:
I really thought that the orcidProfile plugin already did this for the login page, but reading the code I only saw it on the registration page.  I might be missing it elsewhere.

ctgraham commented on templates/optionalShibbolethLogin.tpl:
```diff
@@ -0,0 +1,116 @@
+                {if $shibbolethTitle !== ""}
+                        {$shibbolethTitle}
+                {else}
+                        Institutional Login
```
This text should be in the locale file.

ctgraham commented on templates/optionalShibbolethLogin.tpl:
```diff
@@ -0,0 +1,116 @@
+                        {if $shibbolethButtonLabel !== ""}
+                                {$shibbolethButtonLabel}
+                        {else}
+                                Institutional Login
```
This text should be in the locale file.

ctgraham commented on templates/optionalShibbolethLogin.tpl:
```diff
@@ -0,0 +1,116 @@
+        </a>
+
+        <h2>
+                Local Login
```
This text should be in the locale file.

ctgraham reviewed (changes requested):
I did a quick read-through with some suggestions/questions.  I'll try to schedule some time to install and test the proposed changes.

Commit:
Added Shibboleth Login to Registration Page.
Removed custom template in favour of registration hooks.
Addresses issues found in pull review

Commit:
Removed debugging code. Removed forced forward slash.

Commit:
Added missing templateMgr.

Commit:
Removed force forward slash.

lmeerwood commented:
Hey @ctgraham, I've gone through and made all those changes you requested. I have run out of time unfortunately and could not make the changes requested about the SITE CONTEXT and multiple / single journals. As such I have reverted all the code that was changed to make it work for Single journals. This means issue #3 will no longer be fixed in this merge.

asmecher commented:
@ctgraham and @lmeerwood, are you both satisfied that this is ready to go? It sounds like @davidjb is working with this code and [has given it a thumbs-up](https://github.com/pkp/shibboleth/pull/13#issuecomment-777170008).

ctgraham commented:
The only thing we didn't cover in the earlier code review was clarifying the expectation of the "Shibboleth SP path" setting ("relative url, prefix with slash if you want it to be absolute").  With the .po file conversion, we should adjust that if desired after the merge.

I didn't get a chance to test this in our environment, and won't until sometime in March, so don't let me hold up the merge.  My read-through is good.

davidjb committed:
Merge remote-tracking branch 'origin/master' into shibboleth-optional

This converts the changes from locale.xml into locale.po

davidjb committed:
Fix indentation of _checkAdminStatus

davidjb committed:
Provide examples for SP path setting

davidjb commented:
@ctgraham @asmecher I can take over and make the changes for this PR as needed (I'm on the @jcu-eresearch team so have access to this repo).  I've merged in the other language files changes, updating the en_US .po file in the process.  I've also fixed the indentation issue @ctgraham mentioned.

I'm just catching myself up on the other URL related issue -- the changes about the slash were already reverted so now the WAYF field can have any value the sys admin wants -- be it an absolute URL, absolute path or relative path.  I've highlighted this in ff1ce2f6854b8d93f5a06256bd712d26a3ff53e5.

The current codebase (as of this PR) is deployed and working in our multi-journal 3.3.0 environment.  Login via Shibboleth are option and  work at the site level and the journal level, with a single Shibboleth SP set up running at `/Shibboleth.sso`; local logins work as well.

ctgraham commented:
Didn't realize I had write access to this repo!  Merged, thanks @lmeerwood and @davidjb !

ctgraham commented:
@asmecher , I do not believe I have permissions to package a new release on this repo.

asmecher commented:
@ctgraham, could you try again? I've bumped you up to admin.
